### PR TITLE
development flow for kubecf

### PR DESCRIPTION
**Pull request by andreas-kupries**
## Description

https://jira.suse.com/browse/CAP-522

## Test plan

These are doc changes. There are no automatic tests. (Could make linting of doc syntax automatic). Out of scope for this PR).

### CODE REVIEW

File: doc/mok_brc.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+## CF operator
+
+The [CF operator] is the underlying generic tool to deploy a (modified)
+BOSH deployment like Kubecf for use.
+
+[CF operator]: https://github.com/cloudfoundry-incubator/cf-operator
+
+It has to be installed in the same kube cluster Kubecf will be deployed to.
+
+For simplicity it will be installed from a released helm chart:
```
Comment (f0rmiga) [inline doc/mok_brc.md]: Missing comma after `simplicity`.

File: doc/mok_brc.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+|---		|---				|
+|Deployment	| `bazel run //dev/scf:apply`	|
+|Tear down	| `bazel run //dev/scf:delete`	|
+
+In this default deployment kubecf is launched without Ingress, and
+uses the Diego scheduler.
+
+### Access
+
+To access the clister after the CF operator has completed the
```
Comment (f0rmiga) [inline doc/mok_brc.md]: clister -> cluster.

File: doc/Contribute.md [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+
+  - [Deployment](#deployment)
+  - [Pull Requests](#pull-requests)
+  - [Source Organization](#source-organization)
+  - [Linting](#linting)
+  - [Patching](#patching)
+
+## Deployment
+
+Kubecf is build on top of a number of technologies, namely Kubernetes,
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply doc/Contribute.md]: Fixed with commit 67d8d8687466e4ea2f63516e07cec3002992b8f0.

File: doc/Contribute.md [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+Helm (charts), and the CF operator (for the translation of BOSH
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply doc/Contribute.md]: Fixed with commit 67d8d8687466e4ea2f63516e07cec3002992b8f0.

File: doc/Contribute.md [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+deployments into Kubernetes objects).
+
+For all these we have multiple choices for installing them, and
+various interactions between the choices influence the details of the
+commands to use.
+
+Instead of trying to document all the possibilities and all their
+interactions at once supporting documents will describe specific
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply doc/Contribute.md]: Fixed with commit 67d8d8687466e4ea2f63516e07cec3002992b8f0.

File: doc/Contribute.md [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+combinations of choices in detail, from the bottom up.
+
+|Document		|Description						|
+|---			|---	   						|
+|[MOK_BRC](mok_brc.md)	|Minikube/Bazel + Operator/Release + Kubecf/Checkout	|
+
+## Pull Requests
+
+The general work flow for pull requests contributing bug fixes,
+features, etc. is
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply doc/Contribute.md]: Fixed with commit 67d8d8687466e4ea2f63516e07cec3002992b8f0.

File: doc/Contribute.md [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+  - Implement the bug fix, feature, etc. on that branch/fork.
+
+  - Submit a pull request based on the branch/fork through the github
+    web interface, against the __master__ branch.
+
+  - Developers will review the content of the pull request, asking
+    questions, requesting changes, and generally discussing the
+    submission with the submitter and among themselves.
+
+  - After all issues with the request are resolved, and CI has passed
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply doc/Contribute.md]: Fixed with commit 67d8d8687466e4ea2f63516e07cec3002992b8f0.

File: doc/Contribute.md [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+are:
+
+|Directory			|Content						|
+|---				|---							|
+|__top__			|Documentation entrypoint, License,			|
+|				|Main workspace definitions				|
+|__top__/.../README.md		|Directory-specific local documentation			|
+|__top__/bosh/releases		|Support for runtime patches of a kubecf deployment	|
+|__top__/doc			|Global documentation					|
+|__top__/dev/cf_deployment/bump |Tools to support updating the cf deployment		|
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply doc/Contribute.md]: Fixed with commit 67d8d8687466e4ea2f63516e07cec3002992b8f0.
(Was some left-over tabbing, I believe0

File: doc/Contribute.md [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+    Such a rebase will be a change request from the developers to the
+    contributor, on the assumption that the contributor is best suited
+    to resolving the conflicts.
+
+## Source Organization
+
+The important directories of the kubecf sources, and their contents
+are:
+
+|Directory			|Content						|
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply doc/Contribute.md]: Fixed with commit 67d8d8687466e4ea2f63516e07cec3002992b8f0.

File: doc/Contribute.md [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+report any issues found.
+
+## Patching
+
+### Background
+
+The main goal of the CF operator is to take a BOSH deployment
+manifest, deploy it, and have it run as-is.
+
+Naturally in practice this goal is not quite reached yet, requiring
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply doc/Contribute.md]: Fixed with commit 67d8d8687466e4ea2f63516e07cec3002992b8f0.

File: doc/Contribute.md [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+patching of the deployment manifest in question, and/or the of
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply doc/Contribute.md]: Fixed with commit 67d8d8687466e4ea2f63516e07cec3002992b8f0.

File: doc/Contribute.md [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+involved releases, at various points of the deployment process. The
+reason behind a patch is generally fixing a problem, whether it be
+from the translation into the kube environment, an issue with an
+underlying component, or something else.
+
+Then there are features, given the user of the helm chart wrapped
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply doc/Contribute.md]: Fixed with commit 67d8d8687466e4ea2f63516e07cec3002992b8f0.

File: doc/Contribute.md [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+around the deployment manifest the ability to easily toggle various
+preset configurations, like for example the use of eirini instead of
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply doc/Contribute.md]: Fixed with commit 67d8d8687466e4ea2f63516e07cec3002992b8f0.

File: doc/Contribute.md [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+diego as the application scheduler, etc.
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply doc/Contribute.md]: Fixed with commit 67d8d8687466e4ea2f63516e07cec3002992b8f0.

File: doc/Contribute.md [hunk 15]
```diff
@@ -0,0 +1,229 @@
...
+
+### Features
+
+A features of kubecf is usually implemented using a combination of
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply doc/Contribute.md]: Fixed with commit 67d8d8687466e4ea2f63516e07cec3002992b8f0.

File: doc/Contribute.md [hunk 16]
```diff
@@ -0,0 +1,229 @@
...
+[Helm templating] and [BOSH ops files].
+
+[Helm templating]: https://helm.sh/docs/chart_template_guide/
+[BOSH ops files]:  https://bosh.io/docs/cli-ops-files/
+
+The helm templating is used to translate from the properties in the
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply doc/Contribute.md]: Fixed with commit 67d8d8687466e4ea2f63516e07cec3002992b8f0.

File: doc/Contribute.md [hunk 17]
```diff
@@ -0,0 +1,229 @@
...
+etc.)
+
+The helm templating is applied when the kubecf chart is deployed.
+
+The ops files are then applied by the operator, transforming the base
+manifest from the chart into the final manifest to deploy.
+
+### Customization
+
+For customization during development (and maybe by operators ?) kubecf
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply doc/Contribute.md]: Fixed with commit 67d8d8687466e4ea2f63516e07cec3002992b8f0.

File: doc/Contribute.md [hunk 18]
```diff
@@ -0,0 +1,229 @@
...
+     For example, `kubectl apply` the object below
+
+         ---
+         apiVersion: v1
+         kind: ConfigMap
+         metadata:
+           name: configmap_name
+         data:
+           ops: |-
+               some_random_ops
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply doc/Contribute.md]: Fixed with commit 67d8d8687466e4ea2f63516e07cec3002992b8f0.

File: doc/Contribute.md [hunk 19]
```diff
@@ -0,0 +1,229 @@
...
+
+     and then use `--set .Values.operations.custom[0]=configmap_name`
+     or equivalent as part of a kubecf deployment to include that ops
+     file in the deployment.
+
+  1. The second mechanism allows the specification of any custom BOSH
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply doc/Contribute.md]: This is markdown, which auto-counts in the rendered result. Still fixed.
Fixed with commit 67d8d8687466e4ea2f63516e07cec3002992b8f0.

File: doc/Contribute.md [hunk 20]
```diff
@@ -0,0 +1,229 @@
...
+
+Both forms of customization assume a great deal of familiarity on the
+part of the developer and/or operator with the BOSH releases, instance
+groups and jobs underlying the CF deployment manifest, i.e. which
+properties exist, what changes to them mean and how they affect the
+system.
+
+### Patches
+
+In SCF v1, the predecessor to kubecf, the [patches] scripts enabled
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply doc/Contribute.md]: Fixed with commit 67d8d8687466e4ea2f63516e07cec3002992b8f0.

File: doc/Contribute.md [hunk 21]
```diff
@@ -0,0 +1,229 @@
...
+developers and maintainers to apply general patches to the sources of
+a job (i.e. configuration templates, script sources, etc.) before that
+job was rendered and then executed. At the core the feature allows the
+user to execute custom scripts during runtime of the job container for
+a specific instance_group
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply doc/Contribute.md]: Fixed with commit 67d8d8687466e4ea2f63516e07cec3002992b8f0.

File: doc/Contribute.md [hunk 22]
```diff
@@ -0,0 +1,229 @@
...
+
+[patches]: https://github.com/SUSE/scf/tree/develop/container-host-files/etc/scf/config/scripts/patches
+[Pre render scripts]: https://github.com/cloudfoundry-incubator/cf-operator/blob/master/docs/from_bosh_to_kube.md#Pre_render_scripts
+
+Kubecf makes use of this feature to fix a number of issues in the
+deployment. The relevant patch scripts are found under the directory
+`bosh/releases/pre_render_scripts`.
+
+When following the directory structure explained by the
+[README](../bosh/releases/pre_render_scripts/README.md) the bazel
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply doc/Contribute.md]: Fixed with commit 67d8d8687466e4ea2f63516e07cec3002992b8f0.

File: doc/helm.md [hunk 23]
```diff
@@ -0,0 +1,30 @@
+# Helm
+
+Install Helm with RBAC.
+Note that this requires creating the role first:
+
+```
+kubectl create -f <( cat <<EOF
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply doc/helm.md]: Fixed with commit 67d8d8687466e4ea2f63516e07cec3002992b8f0.

File: doc/mok_brc.md [hunk 24]
```diff
@@ -0,0 +1,126 @@
+# Kubecf deployment via Minikube/Bazel + Operator/Release + Kubecf/Checkout
+
+The intended audience of this document are developers wishing to
+contribute to the Kubecf project.
+
+Here we explain how to deploy Kubecf locally using
+
+  - Minikube to manage a local kubernetes cluster.
+  - Helm.
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply doc/mok_brc.md]: Fixed with commit 67d8d8687466e4ea2f63516e07cec3002992b8f0.

File: doc/mok_brc.md [hunk 25]
```diff
@@ -0,0 +1,126 @@
+# Kubecf deployment via Minikube/Bazel + Operator/Release + Kubecf/Checkout
+
+The intended audience of this document are developers wishing to
+contribute to the Kubecf project.
+
+Here we explain how to deploy Kubecf locally using
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply doc/mok_brc.md]: Fixed with commit 67d8d8687466e4ea2f63516e07cec3002992b8f0.

File: doc/mok_brc.md (hunk 1 above)
Comment (andreas-kupries) [reply doc/mok_brc.md]: Fixed with commit 67d8d8687466e4ea2f63516e07cec3002992b8f0.

File: doc/mok_brc.md (hunk 2 above)
Comment (andreas-kupries) [reply doc/mok_brc.md]: Fixed with commit 67d8d8687466e4ea2f63516e07cec3002992b8f0.

File: doc/Contribute.md [hunk 26]
```diff
@@ -0,0 +1,229 @@
...
+groups and jobs underlying the CF deployment manifest, i.e. which
+properties exist, what changes to them mean and how they affect the
+system.
+
+### Patches
+
+In SCF v1, the predecessor to kubecf, the [patches] scripts enabled
+developers and maintainers to apply general patches to the sources of
+a job (i.e. configuration templates, script sources, etc.) before that
+job was rendered and then executed. At the core the feature allows the
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply doc/Contribute.md]: Fixed with commit 67d8d8687466e4ea2f63516e07cec3002992b8f0.

File: doc/Contribute.md [hunk 27]
```diff
@@ -0,0 +1,229 @@
...
+     There is no tooling to help the writer with the ensuing quoting
+     hell.
+
+     Note further that the resulting config maps have to be applied,
+     i.e. uploaded into the kube cluster __before__ deploying the
+     kubecf helm chart with its modified values.yaml.
+
+     For example, `kubectl apply` the object below
+
+         ---
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply doc/Contribute.md]: As-is it is a proper codeblock. Using __```__ should not make a difference.
Using the ```yaml ... Chokes the pandoc I use to preview (nroff/man, and pdf outputs).
Actually, not choking per se. It simply does not recognize ```yaml as a valid codeblock intro and treats it as part of a plain codeblock.

Done with commit 6b2a5ca4303062f23284545b06eaeac61d241b99.

File: doc/Contribute.md [hunk 28]
```diff
@@ -0,0 +1,229 @@
...
+         apiVersion: v1
+         kind: ConfigMap
+         metadata:
+           name: configmap_name
+         data:
+           ops: |-
+               some_random_ops
+
+     and then use `--set .Values.operations.custom[0]=configmap_name`
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply doc/Contribute.md]: Done with commit 92d86dea5f07a621d0b7c8e6e0791f3f6c4fe175.

File: doc/Contribute.md [hunk 29]
```diff
@@ -0,0 +1,229 @@
+# Guide To Contribution
+
+The intended audience of this document are developers wishing to
+contribute to the Kubecf project.
+
+It provides a basic overview of various aspects of the project, from
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply doc/Contribute.md]: Done with commit a65a8143ec4daf7e7ac68d209b456cb778b68c59.

File: doc/Contribute.md (hunk 27 above)
[the remark this answers is not in the document]
Comment (f0rmiga) [reply doc/Contribute.md]: Hmm interesting! I didn't know.

**Pull request merged**